### PR TITLE
top-p/nucleus sampling for greedy search

### PR DESCRIPTION
**Pull request by funboarder13920**
todo:
- [x] benchmark inference speed CPU/GPU  + benchmark old vs new implementation speed
- [x] extensive GreedySearch tests for top-p sampling, beam_size and other new options
- [x] remove `<unk>` token option
- [x] keep `<eos>` in sampling option 
- [x] length penalty for greedy search
- [x] extend GreedySearch to beam_size > 1 ?

Benchmark results on gpt2 medium :
- initial VS #1958 implementation : 70 tok/s vs 69.5 tok/s (1 beam, topk 50, 100 tokens generated per sequence)
- huggingface VS #1958 implementation :  47.5 tok/s vs 47.0 tok/s (10 beams, topk 50, topp 0.95, 100 tokens generated per sequence)

### CODE REVIEW

File: onmt/opts.py [hunk 1]
```diff
@@ -658,6 +658,12 @@ def _add_decoding_opts(parser):
                    "sampling restricted to the k most likely next tokens. "
```
Comment (funboarder13920) [inline onmt/opts.py:658]: random_sampling_topk defaulting to 1 will make random_sampling_top_p useless.
I can:
- default it to -1 or 0
- make an assertion when parsing the config : Assert random_sampling_top_p == 0 or random_sampling_topk != 1

**Comment by funboarder13920** [comment]
Tests will be fixed by #1960

File: data/data_lm/gen-sampling-beams-sol.txt [hunk 2]
```diff
@@ -0,0 +1,7 @@
+automatically ?
```
Comment (funboarder13920) [inline data/data_lm/gen-sampling-beams-sol.txt:1]: no length penalty / no normalization : shortest sequence is selected ?

Comment (funboarder13920) [reply data/data_lm/gen-sampling-beams-sol.txt:1]: GlobalScorer will be added to greedy search, length penalty only for now

File: onmt/opts.py [hunk 3]
```diff
@@ -627,6 +627,16 @@ def train_opts(parser):
 
 
 def _add_decoding_opts(parser):
+    group = parser.add_argument_group('Decoding')
+    group.add('--decoding_strategy', '-decoding_strategy',
```
[the remark this answers is not in the document]
Comment (Zenglinxiao) [reply onmt/opts.py:631]: `random_sampling_topk` will only take effect when `beam_size` is set to 1 before this change, it's not automatically activated by default. People using it is turning off beam search and use topk explicitly, I think it's safe to change this.

**Comment by funboarder13920** [comment]
Tests fixed in #1972

File: onmt/translate/greedy_search.py [hunk 4]
```diff
@@ -78,38 +123,44 @@ class GreedySearch(DecodeStrategy):
...
     """
 
-    def __init__(self, pad, bos, eos, batch_size, min_length,
-                 block_ngram_repeat, exclusion_tokens, return_attention,
-                 max_length, sampling_temp, keep_topk):
+    def __init__(self, pad, bos, eos, unk, batch_size, global_scorer,
+                 min_length, block_ngram_repeat, exclusion_tokens,
+                 return_attention, max_length, sampling_temp, keep_topk,
+                 keep_topp, beam_size, ban_unk_token, always_sample_eos):
         assert block_ngram_repeat == 0
```
Comment (funboarder13920) [inline onmt/translate/greedy_search.py:135]: @Zenglinxiao do you recall why this assertion exists ? It seems like blocking ngrams is supported by GreedySearch

Comment (Zenglinxiao) [reply onmt/translate/greedy_search.py:135]: At the time of that PR, `block_ngram_repeat` has not properly implemented yet. I think we can remove this assertion now.

File: .github/workflows/push.yml [hunk 5]
```diff
@@ -232,10 +232,39 @@ jobs:
...
+        diff data/data_lm/gen-nucleus-sampling-sol.txt /tmp/gen && rm /tmp/gen
+    - name: Test LM generation with random sampling multi-beams
+      run: |
+        python translate.py -model onmt/tests/test_model_lm.pt \
+          -src data/data_lm/src-gen.txt \
+          -verbose -batch_size 10 \
+          -beam_size 10 \
+          -seed 1 \
+          -random_sampling_topk 50 \
+          -random_sampling_topp 0.95 \
```
Comment (francoishernandez) [inline .github/workflows/push.yml:260]: Is it intended to have both `-random_sampling_topk` and `-random_sampling_topp` set here?

File: onmt/opts.py [hunk 6]
```diff
@@ -643,30 +644,42 @@ def _add_decoding_opts(parser):
                    "target token. If it is not provided (or the identified "
                    "source token does not exist in the table), then it "
                    "will copy the source token.")
+    group.add('--ban_unk_token', '-ban_unk_token',
+              action="store_true",
+              help="Prevent unk token use by setting unk proba to 0")
```
Comment (francoishernandez) [inline onmt/opts.py:649]: Maybe "generation" instead of "use", the unk token will still be used on source side if a token is not in vocab.

File: onmt/opts.py [hunk 7]
```diff
@@ -643,30 +644,42 @@ def _add_decoding_opts(parser):
...
     group.add('--random_sampling_temp', '-random_sampling_temp',
               default=1., type=float,
               help="If doing random sampling, divide the logits by "
                    "this before computing softmax during decoding.")
+    group.add('--always_sample_eos', '-always_sample_eos',
+              action="store_true",
+              help="Whether to keep eos among sampled from token list")
     _add_reproducibility_opts(parser)
 
     group = parser.add_argument_group('Beam Search')
```
Comment (francoishernandez) [inline onmt/opts.py:682]: Keeping the "Beam Search" group but moving `beam_size` somewhere else might be a bit misleading.

File: onmt/opts.py [hunk 8]
```diff
@@ -643,30 +644,42 @@ def _add_decoding_opts(parser):
...
+    group.add('--ban_unk_token', '-ban_unk_token',
+              action="store_true",
+              help="Prevent unk token use by setting unk proba to 0")
     group.add('--phrase_table', '-phrase_table', type=str, default="",
               help="If phrase_table is provided (with replace_unk), it will "
                    "look up the identified source token and give the "
                    "corresponding target token. If it is not provided "
                    "(or the identified source token does not exist in "
                    "the table), then it will copy the source token.")
+    group.add('--beam_size', '-beam_size', type=int, default=5,
```
Comment (francoishernandez) [inline onmt/opts.py:656]: Not sure why this was moved here.

File: onmt/translate/greedy_search.py [hunk 9]
```diff
@@ -38,17 +87,10 @@ def sample_with_temperature(logits, sampling_temp, keep_topk):
             topk_scores /= sampling_temp
     else:
         logits = torch.div(logits, sampling_temp)
-
+        if keep_topp > 0:
+            logits = sample_topp(logits, keep_topp, always_sample_eos, eos)
         if keep_topk > 0:
```
Comment (francoishernandez) [inline onmt/translate/greedy_search.py:92]: I don't find it very clear how topp and topk intertwine. Does the original paper intend for both to be used together, and in that order?

File: onmt/opts.py [hunk 10]
```diff
@@ -643,30 +644,42 @@ def _add_decoding_opts(parser):
...
+              default=0, type=float,
+              help="Probability for top-p/nucleus sampling. Restrict tokens"
+                   " to the most likely until the cumulated probability is"
+                   " over p. In range [0, 1]."
+                   " https://arxiv.org/abs/1904.09751")
     group.add('--random_sampling_temp', '-random_sampling_temp',
               default=1., type=float,
               help="If doing random sampling, divide the logits by "
                    "this before computing softmax during decoding.")
+    group.add('--always_sample_eos', '-always_sample_eos',
```
Comment (francoishernandez) [inline onmt/opts.py:677]: What are the cases where we don't want to sample eos? Shouldn't this be handled via `min_length` instead?

File: .github/workflows/push.yml (hunk 5 above)
Comment (funboarder13920) [reply .github/workflows/push.yml:260]: Yes it is intended, both can be used together

File: onmt/translate/greedy_search.py (hunk 9 above)
Comment (funboarder13920) [reply onmt/translate/greedy_search.py:92]: The order won't change the result, it's like a logical "and".
The original paper does not combine the two methods.
Huggingface allows both methods to be combined. With fairseq it is one or the other.

For example if after topp filtering there is more than k tokens selected, topk will limit the amount of tokens to k. It's a second filtering step.
You can always use only one of the two, it is not necessary to use both at the same time

File: onmt/opts.py (hunk 10 above)
Comment (funboarder13920) [reply onmt/opts.py:677]: This is specific to topk/topp. Even if eos token is not in the topk/topp tokens, we keep it rather than excluding it.
If there is a min_length, eos will still be removed, always_sample_eos does not overrule min_length.

File: onmt/opts.py (hunk 8 above)
Comment (funboarder13920) [reply onmt/opts.py:656]: I used the beam_size option in greedy search because we can generate several examples in one pass with sampling too. Hence this argument is not beam search specific anymore.
I can't duplicate the option.
I can create another option param specific to sampling but it will be used exactly like beam_size and will be named beam_size in shared part of the code as well like DecodeStrategy

Comment (francoishernandez) [reply onmt/opts.py:656]: Ok, makes sense. Maybe just put a short comment here for those who might be asking similar questions.

File: onmt/opts.py (hunk 10 above)
Comment (funboarder13920) [reply onmt/opts.py:677]: I will remove it, I don't think we will really need this option

File: onmt/opts.py (hunk 8 above)
Comment (funboarder13920) [reply onmt/opts.py]: I found a way of showing that the parameter is shared in the code and in the -h, let me now what you think about it https://github.com/OpenNMT/OpenNMT-py/pull/1958/commits/94cfe91039581a6b67bde79546b768e517caaa0c#diff-c7951b1c7894cf513192baf6a05cebd25719e465c9edef24b102570877aec3a1R686
